### PR TITLE
Project/release 1.0 54 metric dict

### PR DESCRIPTION
**Pull request by mitzimorris**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

Adds logic to allow users to specify custom metric (inv diagnoal or dense covariance metric of the unconstrained parameters) as either a Python dict containing entry 'inv_metric" or list of per-chain dicts, with corresponding unit tests.

Also did the following general code cleanups:

- ran files `cmdstan_arg.py' and 'model.py' through spell-checker to catch typos in both docstrings and error messages.
- corrected error messages in file cmdstan_args.py so that if error message is a proper sentence, first word is capitalized and ends with a period.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by mitzimorris** [comment]
@WardBrian - when you get a chance - mypy isn't happy.  not sure why.  need your help here.

File: cmdstanpy/cmdstan_args.py [hunk 1]
```diff
@@ -104,139 +114,199 @@ def validate(self, chains: Optional[int]) -> None:
         if self.iter_warmup is not None:
             if self.iter_warmup < 0 or not isinstance(self.iter_warmup, int):
                 raise ValueError(
-                    'iter_warmup must be a non-negative integer,'
-                    ' found {}'.format(self.iter_warmup)
+                    'Vaule for iter_warmup must be a non-negative integer,'
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:117]:
```suggestion
                    'Value for iter_warmup must be a non-negative integer,'
```

File: cmdstanpy/cmdstan_args.py [hunk 2]
```diff
@@ -104,139 +114,199 @@ def validate(self, chains: Optional[int]) -> None:
...
-                    )
-                for i, metric in enumerate(self.metric):
-                    if not os.path.exists(metric):
-                        raise ValueError('no such file {}'.format(metric))
-                    if i == 0:
-                        dims = read_metric(metric)
-                    else:
-                        dims2 = read_metric(metric)
-                        if len(dims) != len(dims2):
+                if all(isinstance(elem, Dict) for elem in self.metric):
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:206]:
```suggestion
                if all(isinstance(elem, dict) for elem in self.metric):
```

File: cmdstanpy/cmdstan_args.py [hunk 3]
```diff
@@ -104,139 +114,199 @@ def validate(self, chains: Optional[int]) -> None:
...
+                    metric_files: List[str] = []
+                    for i, metric in enumerate(self.metric):
+                        metric_dict: Union[str, Dict[str, Any]] = self.metric[i]
+                        if 'inv_metric' not in metric_dict:
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:210]: When metric_dict could be a str?

File: cmdstanpy/cmdstan_args.py [hunk 4]
```diff
@@ -104,139 +114,199 @@ def validate(self, chains: Optional[int]) -> None:
...
-                    if not os.path.exists(metric):
-                        raise ValueError('no such file {}'.format(metric))
-                    if i == 0:
-                        dims = read_metric(metric)
-                    else:
-                        dims2 = read_metric(metric)
-                        if len(dims) != len(dims2):
+                if all(isinstance(elem, Dict) for elem in self.metric):
+                    metric_files: List[str] = []
+                    for i, metric in enumerate(self.metric):
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:208]: If we move this loop up one block and check dict / str for each element, we could then also accept user to give a combination of dict and str

File: cmdstanpy/cmdstan_args.py [hunk 5]
```diff
@@ -790,36 +859,34 @@ def validate(self) -> None:
             if not os.path.exists(self.data):
                 raise ValueError('no such file {}'.format(self.data))
         elif self.data is not None and not isinstance(self.data, (str, dict)):
-            raise ValueError('data must be string or dict')
+            raise ValueError('Arugment "data" must be string or dict')
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:862]:
```suggestion
            raise ValueError('Argument "data" must be string or dict')
```

File: cmdstanpy/cmdstan_args.py [hunk 6]
```diff
@@ -104,139 +114,199 @@ def validate(self, chains: Optional[int]) -> None:
...
+                            dims = list(
+                                np.asarray(metric_dict['inv_metric']).shape
+                            )
+                        else:
+                            dims2 = list(
+                                np.asarray(metric_dict['inv_metric']).shape
                             )
-                        for dim, dim2 in zip(dims, dims2):
-                            if dim != dim2:
+                            if not (dims == dims2):
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:223]:
```suggestion
                            if dims != dims2:
```

File: cmdstanpy/cmdstan_args.py [hunk 7]
```diff
@@ -104,139 +114,199 @@ def validate(self, chains: Optional[int]) -> None:
...
                                 )
-            if any(dims):
-                if len(dims) > 2 or (len(dims) == 2 and dims[0] != dims[1]):
-                    raise ValueError('bad metric specifiation')
-                self.metric_file = self.metric
-                if len(dims) == 1:
-                    self.metric = 'diag_e'
-                elif len(dims) == 2:
-                    self.metric = 'dense_e'
+                            if not (dims == dims2):
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:255]:
```suggestion
                            if dims != dims2:
```

File: cmdstanpy/cmdstan_args.py [hunk 8]
```diff
@@ -790,36 +859,34 @@ def validate(self) -> None:
...
                 if len(self.inits) != len(self.chain_ids):
                     raise ValueError(
-                        'number of inits files must match number of chains,'
-                        ' found {} inits files for {} chains '.format(
+                        'Number of inits files must match number of chains,'
+                        ' found {} inits files for {} chains.'.format(
                             len(self.inits), len(self.chain_ids)
                         )
                     )
                 names_set = set(self.inits)
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:889]:
```suggestion

```

File: cmdstanpy/cmdstan_args.py [hunk 9]
```diff
@@ -2,6 +2,7 @@
 CmdStan arguments
 """
 import logging
+import numpy as np
```
Comment (ahartikainen) [inline cmdstanpy/cmdstan_args.py:5]: Run isort once?

File: test/test_cmdstan_args.py [hunk 10]
```diff
@@ -509,7 +505,7 @@ def test_args_bad(self):
         jinits2 = os.path.join(DATAFILES_PATH, 'bernoulli.init_2.json')
```
Comment (ahartikainen) [inline test/test_cmdstan_args.py:505]:
```suggestion

```

Same for jinits1

**Review by ahartikainen — COMMENTED** [review]
Added some comments

File: cmdstanpy/cmdstan_args.py (hunk 4 above)
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:208]: we could, but should we?  I don't see a good use case.

Comment (ahartikainen) [reply cmdstanpy/cmdstan_args.py:208]: True, probably causes more damage :D

File: cmdstanpy/cmdstan_args.py (hunk 1 above)
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:117]: thanks! spell-checked the entire file - fixed several typos.

File: cmdstanpy/cmdstan_args.py (hunk 3 above)
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:210]: never?  good catch!

File: cmdstanpy/cmdstan_args.py (hunk 6 above)
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:223]: of course - done!

File: cmdstanpy/cmdstan_args.py (hunk 8 above)
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:889]: good catch!

File: cmdstanpy/cmdstan_args.py (hunk 9 above)
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:5]: cool!

File: cmdstanpy/cmdstan_args.py [hunk 11]
```diff
@@ -104,139 +111,204 @@ def validate(self, chains: Optional[int]) -> None:
...
+                        'Expecting {} per-chain step_size specifications, '
+                        ' found {}.'.format(chains, len(self.step_size))
                     )
-                for step_size in self.step_size:
+                for i, step_size in enumerate(self.step_size):
                     if step_size < 0:
                         raise ValueError(
-                            'step_size must be > 0, found {}'.format(step_size)
+                            'Argument "step_size" must be > 0, '
+                            'found {} at index {}'.format(step_size, i)
```
[the remark this answers is not in the document]
Comment (mitzimorris) [reply cmdstanpy/cmdstan_args.py:161]: good call - will do

**Pull request merged**